core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -992,12 +992,20 @@ private[spark] class BlockManager(
 
   /**
    * Release a lock on the given block with explicit TID.
-   * The param `taskAttemptId` should be passed in case we can't get the correct TID from
-   * TaskContext, for example, the input iterator of a cached RDD iterates to the end in a child
+   * The param `taskContext` should be passed in case we can't get the correct TaskContext
+   * for example, the input iterator of a cached RDD iterates to the end in a child
    * thread.
    */
-  def releaseLock(blockId: BlockId, taskAttemptId: Option[Long] = None): Unit = {
-    blockInfoManager.unlock(blockId, taskAttemptId)
+  def releaseLock(blockId: BlockId, taskContext: Option[TaskContext] = None): Unit = {
+    val taskAttemptId = taskContext.map(_.taskAttemptId())
+    // SPARK-27666. Child thread spawned from task thread could produce race condition
```

Comment:
A simpler explanation: When a task completes, Spark automatically releases all the blocks locked by this task. We should not release any blocks for a task that is already completed.